modules/widespaceBidAdapter.js

```diff
@@ -0,0 +1,239 @@
+import { version } from '../package.json';
+import { config } from 'src/config';
+import { registerBidder } from 'src/adapters/bidderFactory';
+import {
+  cookiesAreEnabled,
+  parseQueryStringParameters,
+  parseSizesInput,
+  getTopWindowReferrer
+} from 'src/utils';
+
+const BIDDER_CODE = 'widespace';
+const WS_ADAPTER_VERSION = '2.0.0';
+const LOCAL_STORAGE_AVAILABLE = window.localStorage;
+const COOKIE_ENABLED = cookiesAreEnabled();
+const LS_KEYS = {
+  PERF_DATA: 'wsPerfData',
+  LC_UID: 'wsLcuid',
+  CUST_DATA: 'wsCustomData'
+};
+
+let preReqTime = 0;
+
+export const spec = {
+  code: BIDDER_CODE,
+
+  supportedMediaTypes: ['banner'],
+
+  isBidRequestValid: function(bid) {
+    if (bid.params && bid.params.sid) {
+      return true;
+    }
+    return false;
+  },
+
+  buildRequests: function(validBidRequests) {
+    let serverRequests = [];
+    const REQUEST_SERVER_URL = getEngineUrl();
+    const DEMO_DATA_PARAMS = ['gender', 'country', 'region', 'postal', 'city', 'yob'];
+    const PERF_DATA = getData(LS_KEYS.PERF_DATA).map(perf_data => JSON.parse(perf_data));
+    const CUST_DATA = getData(LS_KEYS.CUST_DATA, false)[0];
```

Comment:
I don't see a storeData call for CUST_DATA so wouldn't it always be empty

Reply:
This will be set by our server and I will just pass it on in next request when available

Reply:
I couldn't see how it ever gets set to local storage.  Am I just missing where that happens?

Reply:
No, you are not missing anything because `CUST_DATA` is not `set` by adapter but rather by  adCreative code. This allows us get some sort of feedback when our won bid creative is rendered. Example: - If we detected some problem when trying to render ad i.e. unable to measure inscreen for some reason ? we will store the reason. - Get device information i.e. fine grained device resolution for iPhone models which not available in user agent HTTP header, it helps us to deliver better creative next time.